collector/export/log_fluent.go

```diff
@@ -0,0 +1,152 @@
+		if !ok {
+			continue
+		}
+		batchToSend = append(batchToSend, log)
+	}
+	if len(batchToSend) == 0 {
+		return nil, nil
+	}
+
+	// Sort based on tags. This allows us to create a message per tag.
+	slices.SortFunc(batchToSend, func(a, b *types.Log) int {
+		tagOne := a.Attributes[tagAttribute]
+		tagTwo := b.Attributes[tagAttribute]
+		return strings.Compare(tagOne.(string), tagTwo.(string))
+	})
+
+	e.w = e.w[:0]
+
+	var err error
+	activeTag := ""
+	activeTagStart := 0
+	for idx, log := range batchToSend {
+		currTag := log.Attributes[tagAttribute].(string)
+		if idx == 0 {
+			activeTag = currTag
+		}
+
+		if idx == len(batchToSend)-1 {
+			err = e.appendMsg(batchToSend[activeTagStart:], activeTag)
+			if err != nil {
+				return nil, err // skip???
+			}
+			return e.w, nil
+		}
+
+		peekTag := batchToSend[idx+1].Attributes[tagAttribute].(string)
+		if peekTag != activeTag {
+			err = e.appendMsg(batchToSend[activeTagStart:idx+1], activeTag)
+			if err != nil {
+				return nil, err // skip???
+			}
+			activeTag = peekTag
+			activeTagStart = idx + 1
+		}
+	}
+
+	return e.w, nil
+}
+
+func (e *fluentEncoder) appendMsg(batchToSend []*types.Log, tag string) error {
+	// Write Forward mode. See https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#forward-mode
+
+	// We use the []byte oriented API instead of the writer-oriented API from msgp.
+	// the Writer oriented one uses unsafe to slightly speed things up, but we have no reason to use this.
+	// Additionally, the writer api does its own buffering and seems overly complicated for what we need.
+	var err error
+	e.w = msgp.AppendArrayHeader(e.w, 2)                        // [<tag>, <entries>]
+	e.w = msgp.AppendString(e.w, tag)                           // <tag>
+	e.w = msgp.AppendArrayHeader(e.w, uint32(len(batchToSend))) // <entries>
+	for _, log := range batchToSend {
+		e.w = msgp.AppendArrayHeader(e.w, 2) // [<time>, <record>]
+		e.fluentExtTime.UnixTsNano = log.Timestamp
+		e.w, err = msgp.AppendExtension(e.w, e.fluentExtTime) // <time>
+		if err != nil {
+			return err // TODO - skip???
+		}
+		// TODO - also put in resource/attributes?
+		e.w, err = msgp.AppendMapStrIntf(e.w, log.Body)
+		if err != nil {
+			return err // TODO - skip???
+		}
+	}
+	return nil
+}
+
+type fluentExtTime struct {
+	UnixTsNano uint64
+}
+
+func (e *fluentExtTime) ExtensionType() int8 {
+	return 0
+}
+
+func (e *fluentExtTime) Len() int {
+	return 8
+}
+
+// https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#eventtime-ext-format
+func (e *fluentExtTime) MarshalBinaryTo(b []byte) error {
+	seconds := uint32(e.UnixTsNano / 1_000_000_000)
+	nanoSeconds := uint32(e.UnixTsNano % 1_000_000_000)
+	binary.BigEndian.PutUint32(b, uint32(seconds))
+	binary.BigEndian.PutUint32(b[4:], uint32(nanoSeconds))
+	return nil
+}
+
+func (e *fluentExtTime) UnmarshalBinary(b []byte) error {
+	seconds := binary.BigEndian.Uint32(b)
+	nanoSeconds := binary.BigEndian.Uint32(b[4:])
```

Comment:
This could panic on a nil slice or a slice of len() < 4.

Reply:
They document they always provide a large enough slice (based on `Len()`), but the check is very cheap so I'm adding one.